api/src/modules/sourcing-records/sourcing-record.repository.ts

```diff
@@ -1,18 +1,29 @@
-import { EntityRepository, Repository } from 'typeorm';
+import { EntityRepository, Repository, SelectQueryBuilder } from 'typeorm';
 import { SourcingRecord } from 'modules/sourcing-records/sourcing-record.entity';
+import { SourcingLocation } from 'modules/sourcing-locations/sourcing-location.entity';
 
 @EntityRepository(SourcingRecord)
 export class SourcingRecordRepository extends Repository<SourcingRecord> {
-  async getYears(): Promise<number[]> {
-    const sourcingRecordsYears: any[] = await this.createQueryBuilder(
-      'sourcingRecords',
-    )
-      .select('year')
-      .distinct(true)
-      .getRawMany();
+  async getYears(materialIds?: string[]): Promise<number[]> {
+    const queryBuilder: SelectQueryBuilder<SourcingRecord> =
+      this.createQueryBuilder('sr').select('year').distinct(true);
 
-    return sourcingRecordsYears
-      .map((elem: { year: number }) => elem.year)
-      .sort();
+    if (materialIds) {
+      queryBuilder.leftJoin(
+        SourcingLocation,
+        'sl',
+        'sl.id = sr."sourcingLocationId"',
+      );
+      queryBuilder.andWhere('"sl"."materialId" IN (:...materialIds)', {
+        materialIds,
+      });
+    }
+    const sourcingRecordsYears: SourcingRecord[] =
+      await queryBuilder.getRawMany();
+    return [
```

Comment:
That's why I use the set. The DB itself returns repeated values (even with the distinct statement) because of how we join both tables, see that the 'distinct' is present which work as expected when no join is required, but not when it is. 

I've been testing it straight to DB